Tss 2fa

ericamador opened the pull request.

ericamador committed:
Add everything necessary for setting up a CAC 2FA method

This commit adds all of the changes necessary for associating a CAC 2FA
method with a user. At this point you can not log in with this method,
but that will be added in the next commit.

ericamador committed:
Add a Django message when a user creates a new 2FA method

This message appears when as a confirmation to the user whenever a new
2FA method has been created.

ericamador committed:
Copy auth_backend.py from django-x509-auth

This authentication backend file was copied from the django-x509-auth
app at commit bb6810c4d641a76e. It will be modified in a following
commit.

ericamador committed:
Update auth_backend from django-x509-auth

This commit simply updates various strings and model names to match the
changes that have been made since bits have been ported from
django-x509-auth to django-two-factor-auth.

ericamador committed:
Add CAC login support

This commit adds everything necessary to log in using a CAC as a 2FA
device.

ericamador committed:
Use better method of determining what 2FA methods can be used

This is a more flexible manner of allowing each individual site to
specify what 2FA methods can be used via site settings.

ericamador committed:
Comment out a line that was causing a bug upstream

The line that is commented out in this commit causes an exception when
the list of available methods has a length of one. I will fix this
upstream at a later time.

ericamador commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Disable all non 2FA login methods.](https://trello.com/c/E3QUXsSw/2648-disable-all-non-2fa-login-methods)

shaunbrady commented on two_factor/views/core.py:
```diff
@@ -162,6 +169,28 @@ def render(self, form=None, **kwargs):
+                # HTTP headers not set
+                messages.error(
+                    self.request,
+                    'We did not get any certificate information for your '
```
If you're going back.... we could factor this out.  That reeks of me :D

ericamador replied:
Thanks for reminding me of this. Wanted to take it out before merging.

ericamador committed:
Factor out X509 error messages as settings

This is done to increase the ability to change these settings on a
per-site basis and to make the code more DRY.
